lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1398,8 +1398,8 @@ bool TypeChecker::coercePatternToType(Pattern *&P, DeclContext *dc, Type type,
         }
       }
       // Otherwise, see if we can introduce a cast pattern to get from an
-      // existential pattern type to the enum type.
-      else if (type->isAnyExistentialType()) {
+      // existential or archetype pattern type to the enum type.
+      else if (type->isAnyExistentialType() || type->is<ArchetypeType>()) {
```

Comment:
Unrelated to this PR, but should we be allowing cases where the pattern is an existential metatype? Unless I'm mistaken, such a pattern can never be matched by an enum case.